emscripten.py

```diff
@@ -1951,25 +1955,43 @@ def create_asm_consts_wasm(forwarded_json, metadata):
     const = asstr(v[0])
     sigs = v[1]
     const = trim_asm_const_body(const)
-    const = '{ ' + const + ' }'
     args = []
-    arity = max(map(len, sigs)) - 1
+    max_arity = 16
+    arity = 0
+    for i in range(max_arity):
+      if ('$' + str(i)) in const:
+        arity = i + 1
     for i in range(arity):
       args.append('$' + str(i))
-    const = 'function(' + ', '.join(args) + ') ' + const
+    const = 'function(' + ', '.join(args) + ') {' + const + '}'
     asm_consts[int(k)] = const
     all_sigs += sigs
 
   asm_const_funcs = []
   for sig in set(all_sigs):
     forwarded_json['Functions']['libraryFunctions']['_emscripten_asm_const_' + sig] = 1
-    args = ['a%d' % i for i in range(len(sig)-1)]
-    all_args = ['code'] + args
     asm_const_funcs.append(r'''
-function _emscripten_asm_const_%s(%s) {
-return ASM_CONSTS[code](%s);
-}''' % (asstr(sig), ', '.join(all_args), ', '.join(args)))
-
+function _emscripten_asm_const_%s(code, sig_ptr, argbuf) {
+  var sig = UTF8ToString(sig_ptr);
```

Comment:
`AsciiToString` would bring in a lot less support code, and should be ok since it's just the signature?

Even with that, though, this adds a bunch of code for each EM_ASM signature, and extra work at runtime. Both worry me.

Remind me, what's the context here? Is this fixing an urgent and unavoidable problem, or is this a refactoring for some future benefit?

Reply:
This lets us remove prototypeless function declarations in C, which unblocks fixing a wasm backend bug: https://bugs.llvm.org/show_bug.cgi?id=35385 . I'd say not urgent, but also not avoidable.
That approach also breaks for the LLD-using path because we essentially generate invalid wasm,  but the current s2wasm approach works because we can represent the invalid wasm in text. Similarly this isn't urgent, but is something we will need to fix before we can use LLD by default.

To note: this reduces the number of signatures generated from O(callsites), to at most two (`iii` and `dii`). Because all the arguments get lowered into a vararg buffer, i.e. the last `i`.
Which part of the extra runtime work is concerning, the C++ `std::string` part, or the JS-side argument-fishing part? Both?

Reply:
Both sides worry me, sorry to say.

Have we considered changing the EM_ASM api to include the type? I mean,
```
int x = EM_ASM_idd({
  // idd : receive two doubles, return an int
  return ($0 + $1) | 0;
}, 3.14159, 2.18281);
```
If this isn't urgent, maybe we have time to change the API. Sort of like we're thinking of doing with addFunction. But I'm not sure if adding the type can help here, what do you think?

Alternatively, we can wait for EM_JS, or we can just accept that EM_ASM is less efficient in the wasm backend?

Reply:
I'm not sure how that'd help? We'd need it to be `EM_ASM("idd", {...}, 1.23, 4.56)`, which would still have the same string-reading issue.

I'm ok with EM_ASM being slower because EM_JS should take over for most cases. I do want to measure how much slower this is though first.

Reply:
Sorry, I wasn't clear - I meant that literally it would be `EM_ASM_idd`, no string. That is, the header would define
```
#define EM_ASM_idd emscripten_asm_const_idd([etc.])
```
(we'd use a script to autogenerate those).

Yeah, measuring is a good idea, maybe my concerns are overblown.

Reply:
How? My first attempt months ago at a C-macro-only approach was indeed to generate every possible combination of signatures via recursive macros. For supporting up to ten arguments we needed tens of thousands of signatures, which bloated compile time to tens of seconds per C source file. Maybe that was due to preprocessor expansion, but those declarations do need to be kept in IR all the way until we run LTO.

We could also limit the number of arguments to something in the 4-6 range? Then the generative approach can work, and the runtime mechanism would be the same as it is today: calling with an index into the ASM_CONSTS array and a static signature, which would avoid the cost of reading from a vararg buffer.

With https://github.com/kripken/emscripten/pull/6163/commits/95536f87f9e232248f79ab6f692f1eacaebbd496 we're able to avoid `std::string`, so the compiler is able to inline the signature string building, turning it in to some `store`s of stack-allocated memory, so the C++ side is as good as the C side with optimizations enabled.

With `-std=c++14` and constexpr, -O2 is able to convert the signatures into a string literal that gets included in the data section, i.e. at runtime it's just an `i32.const`. In theory we can `#if __cplusplus >= 201402L` the constexpr version, but the last thing this file needs right now is more conditional complexity. Mentioning it here in case we decide it's worth it to do.

Reply:
Yeah, I guess I was assuming in my head that we'd limit to the 4-6 range. That's not great, but if otherwise things are good, maybe it's worth considering?

Reply:
Ok, so, turns out wasm->js overhead is way less than what I was expecting. Which is to say, the extra loads and stores into the vararg buffer adds a ton of overhead.
Compiling this micro-benchmark before and after:
```c++
#include <emscripten.h>
int main() {
  const int NUM_ROUNDS = 10000000;
  const int NUM_PRINTS = 50;
  for (int i = 0; i < NUM_ROUNDS; ++i) {
    int a = EM_ASM_INT(return 5);
    double b = EM_ASM_DOUBLE(return $0 * 2.6, a);
    int c = EM_ASM_INT(return $1 + $0 * $2, i, a, b);
    double d = EM_ASM_DOUBLE(return arguments[$0 % arguments.length],
                             i, a, b, c, i, a, b, c, i);
    if (((i + 1) % (NUM_ROUNDS / NUM_PRINTS)) == 0) {
      EM_ASM(console.log($0), d);
    }
  }
}
```
results in a **10x** slowdown. That's uh, pretty untenable.

So, let's go with breaking the API. Three user-facing options:
A) `EM_ASM_idd(return $0 + $1 * $2, a, b, c)`
B) `EM_ASM(idd, return $0 + $1 * $2, a, b, c)`
C) `EM_ASM(return $0 + $1 * $2, idd, a, b, c)`
B and C are the same with different placement of the argument string.
With A we'll need a bunch of `#define EM_ASM_idd`s, which we'll probably want to generate with a script. I assume we'll want to check in said script, in case we ever need to regenerate that file?
B&C we can do with macros, i.e. `#define EM_ASM(sig, code, ...) emscripten_asm_const_##sig(#code, ##__VA_ARGS__)`, but is maybe harder for users to grok at the callsite.
Which is better?

Reply:
Might make sense to open a new issue for this breaking API change, it's a big topic, and we are in the commit comments section of a PR, that people might miss.

Reply:
Good call, done https://github.com/kripken/emscripten/issues/6213

Reply:
Having thought about this some more, I'm thinking the best thing to do (for now at least) is to just make EM_ASM 10x slower for wasm backend, and if we need to break the API for speed later we're still at liberty to do so.

Reply:
Used `AsciiToString` instead, which saved ~700B (~5%) on that microbenchmark (with `-O3 --closure 1`), and also a ~3% runtime speedup (woohoo, only 9.7x slowdown!)